File: server/feed10/src/outbound.ts
import * as nosql from './nosqldb'
import * as p from './participant'
import * as f from './friend'
import * as i from './inbound'
import { Client } from 'cassandra-driver'
import { SearchService } from './elastic'

export class OutboundModel {
   readonly from: p.ParticipantModel
   readonly occurred: Date
   readonly subject: string
   readonly story: string
   constructor(from: p.ParticipantModel, occurred: Date, subject: string, story: string) {
      this.from = from
      this.occurred = occurred
      this.subject = subject
      this.story = story
   }
}

export class OutboundService extends nosql.Repository {
   readonly participantService: p.ParticipantService
   readonly friendService: f.FriendService
   readonly inboundService: i.InboundService
   readonly searchService: SearchService
   constructor(nosqlClient: Client, ps: p.ParticipantService, fs: f.FriendService, is: i.InboundService, ss: SearchService) {
      super(nosqlClient)
      this.participantService = ps
      this.friendService = fs
      this.inboundService = is
      this.searchService = ss
   }
   public async get(id: number): Promise<OutboundModel[]> {
      const np = new p.ParticipantModel(id, '')
      const query = 'select toTimestamp(occurred) as occurred, subject, story from Outbound where participantid = ? order by occurred desc';
      const results = await this.nosqlClient.execute(query, [id], {prepare: true})
      return results.rows.map((row) => {
      	 return new OutboundModel(np, row.occurred, row.subject, row.story)
      })
   }
   
   public async save(o: OutboundModel): Promise<OutboundModel> {
      const cql = 'insert into Outbound (ParticipantID, Occurred, Subject, Story) values (?, now(), ?, ?)';
      this.nosqlClient.execute(cql, [o.from.id, o.subject, o.story], {prepare: true})
      const friends = await this.friendService.get(o.from.id)
      friends.forEach((friend) => {
         const nim = new i.InboundModel(o.from, friend.to, o.occurred, o.subject, o.story)
	 this.inboundService.save(nim)
      })
      this.searchService.index(o.from.id, o.story)
      return o
   }

   public async search(term: string): Promise<p.ParticipantModel[]> {
      const results = await this.searchService.search(term)
      return Promise.all(results.map(async (id) => {
         return await this.participantService.get(id)
      }))
   }
}

File: client/flutter/lib/models/outbound.dart
import './participant.dart';

class OutboundModel {
  final String occurred;
  final String subject;
  final String story;
  OutboundModel({
    required this.occurred,
    required this.subject,
    required this.story,
  });
}

class OutboundEnvelope {
  List<OutboundModel> build(Map<String, dynamic> payload) {
    final List<OutboundModel> retVal = [];
    payload['data']['outbound'].forEach((item) {
      retVal.add(OutboundModel(
        occurred: item['occurred'],
        subject: item['subject'],
        story: item['story'],
      ));
    });
    return retVal;
  }
}

class SearchResultModel {
  final OutboundModel outbound;
  final ParticipantModel participant;
  SearchResultModel(this.outbound, this.participant);
}

class SearchResultEnvelope {
  List<SearchResultModel> build(Map<String, dynamic> payload) {
    final List<SearchResultModel> retVal = [];
    payload['data']['search'].forEach((item) {
      final o = OutboundModel(
        occurred: item['outbound']['occurred'],
        subject: item['outbound']['subject'],
        story: item['outbound']['story'],
      );
      final p = ParticipantModel(
        id: item['participant']['id'],
	name: item['participant']['name'],
      );
      retVal.add(SearchResultModel(o, p));
    });
    return retVal;
  }
}